# Justin Elkinson
### Worked with Cam Germaine

### Training and tuning a random forest:

In [1]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

In [2]:
# Import the data
df = pd.read_csv('BC Data.csv')

# Assign input variables
X = df.loc[:,['radius_mean','texture_mean', 'perimeter_mean', 'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean', 'concave points_mean',
             'symmetry_mean', 'fractal_dimension_mean',
             'radius_se','texture_se', 'perimeter_se', 'area_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se',
             'symmetry_se', 'fractal_dimension_se']]

# Assign target variable
y = df['diagnosis']


In [3]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.2)

In [4]:
# Create a decision tree and train

rf1 = RandomForestClassifier()
rf1.fit(x_train, y_train)

/usr/local/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [5]:
# Accuracry on training data
rf1.score(x_train, y_train)

0.9912087912087912

In [6]:
# Prediction and accuracy on testing data
y_pred = rf1.predict(x_test)
metrics.accuracy_score(y_pred, y_test)

0.9035087719298246

In [7]:
# Create another decision tree and train
rf2 = RandomForestClassifier(n_estimators=15, max_features=4)
rf2.fit(x_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features=4, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=15,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [8]:
rf2.score(x_train, y_train)

1.0

#### The accuracy of the random forest with 15 trees and number of variables at each split being 4 is: 1.0

In [9]:
y_pred2 = rf2.predict(x_test)
metrics.accuracy_score(y_pred2, y_test)

0.9385964912280702

### Tuning a random Forest:

In [10]:
# Decide the searching range
param_grid = {'n_estimators': range(10,25), 'max_features':range(3,10)}

from sklearn.model_selection import GridSearchCV
rf3 = GridSearchCV(RandomForestClassifier(), param_grid, cv = 5)

In [11]:
rf3.fit(x_train, y_train)

/usr/local/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=None,
                                              oob_score=False,
                                              random_state=None, verbose=0,
                                              warm_start=False),
             iid

In [12]:
rf3.best_params_

{'max_features': 3, 'n_estimators': 24}

In [13]:
y_pred2 = rf3.predict(x_test)
metrics.accuracy_score(y_pred2, y_test)

0.9385964912280702

#### The best hyper parameters are max features = 3 and n_estimators = 24
#### The accuracy of the forest is 93.86%
#### This tuned model is slightly better than the original forest:
#### Orginal Forest = 90.35% | Tuned Forest = 93.68%